contractcourt/htlc_incoming_contest_resolver.go

```diff
@@ -134,28 +134,52 @@ func (h *htlcIncomingContestResolver) Resolve() (ContractResolver, error) {
 	preimageSubscription := h.PreimageDB.SubscribeUpdates()
 	defer preimageSubscription.CancelSubscription()
 
-	// Create a buffered hodl chan to prevent deadlock.
-	hodlChan := make(chan interface{}, 1)
-
-	// Notify registry that we are potentially settling as exit hop
-	// on-chain, so that we will get a hodl event when a corresponding hodl
-	// invoice is settled.
-	event, err := h.Registry.NotifyExitHopHtlc(h.payHash, h.htlcAmt, hodlChan)
-	if err != nil && err != channeldb.ErrInvoiceNotFound {
-		return nil, err
-	}
-	defer h.Registry.HodlUnsubscribeAll(hodlChan)
+	// Define closure to process hodl events either direct or triggered by
+	// later notifcation.
+	processHodlEvent := func(e invoices.HodlEvent) (ContractResolver,
+		error) {
+
+		if e.Preimage == nil {
+			log.Infof("%T(%v): Exit hop HTLC canceled "+
+				"(expiry=%v, height=%v), abandoning", h,
+				h.htlcResolution.ClaimOutpoint,
+				h.htlcExpiry, currentHeight)
+
+			h.resolved = true
```

Comment:
It's a change in behavior to checkpoint and mark the contract has being resolved. It isn't _fully_ resolved here in that it still needs to turn into a regular success resolver.

Reply:
The change in behaviour is that we sometimes don't pull htlcs for invoices for which we do have the preimage. We don't pull if there are not enough blocks left to satisfy the final cltv expiry requirement of the invoice. In exact the same way as we do it in the link.

Suppose we have a hodl invoice with final cltv 100 blocks. An htlc comes in with just 20 blocks left. This htlc would normally be canceled in the link. But if the link goes down after lock-in but before canceling the htlc and we go to chain, we'll end up here in the resolver.

Without this change, we would just mark that invoice as accepted (and notify listeners), even if there were only still 20 blocks left. Sending the accepted event under those conditions would probably violate the caller's assumptions.

Reply:
Another consistency that existed previously: an invoice with final ctlv 20 and an htlc is locked-in with 30 blocks to expiry. In the link the invoice would then be marked settled and the htlc settled too. But if the same thing would happen in a resolver, we would all of a sudden postpone marking the invoice as settled until the final sweep completed. During this time, the sender of the payment is wondering why his payment is stuck.